# CellFindR
scRNAseq tools for clustering written thanks on a basis for Seurat.

## Introduction
CellFindR is a wraparound for many of the functions that exist in Seurat. It it centered around an iterative clustering algorithm that hopes to find unique cell populations. In addition, CellFindR includes many functions to make plotting and data organization much easier and more friendly for the average biologist. 


### Version 1.0.0 08/20/19
published onto github! BioRxiv https://www.biorxiv.org/content/10.1101/739680v1?rss=1!
 

**any inquiries or debugging questions ** can send an email to
 *kyu@ucsf.edu* 
 
 
### Supplementary Files 

## 1. Link to Differential Gene Matrices:
https://ucsf.box.com/s/x4je0w29p4a4fga598jtlxuijt1f5wmg

Table 1: Mouse E14.5 Cochlear Epithelial Floor Expression and Differential Expression. Output from CellFindR for E14.5 cochlear epithelium floor with columns representing log-normalized: mean, standard deviation, average differential expression and corrected p-values for each CellFindR group. Rows represent each measured gene. Blank represents non-significant value.  

Table 2: Mouse Embryonic Epithelium Floor Expression and Differential Expression. Output from CellFindR for batched E12.5 to P2 epithelium floor with columns representing log-normalized: mean, standard deviation, average differential expression and corrected p-values for each CellFindR group. Rows represent each measured gene. Blank represents non-significant value..  

Table 3: Summary of scRNAseq experiments. All experiments labeled with their qualities metrics and cell counts.  

## 2. Link to CellFindR Raw output of all Experiments:
*TBD*

## 3. GEO link to sequencing data
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?&acc=GSE135913
* Will become public once requested* 
